### PR TITLE
Use interface for storage interactions in Appendonly TAM

### PR DESCRIPTION
**Pull request by reshke**
This PR provides basic SMGR infrastructure for AO relations. This interface is needed for Yezzey support.
Developed originally for Greenplum 6, adapted for Cloudberry. 

fix #ISSUE_Number

---

### Change logs

_Describe your change clearly, including what problem is being solved or what feature is being added._

_If it has some breaking backward or forward compatibility, please clary._

### Why are the changes needed?

_Describe why the changes are necessary._

### Does this PR introduce any user-facing change?

_If yes, please clarify the previous behavior and the change this PR proposes._

### How was this patch tested?

_Please detail how the changes were tested, including manual tests and any relevant unit or integration tests._

### Contributor's Checklist

Here are some reminders and checklists before/when submitting your pull request, please check them:

- [x] Make sure your Pull Request has a clear title and commit message. You can take [git-commit](https://github.com/cloudberrydb/cloudberrydb/blob/main/.gitmessage) template as a reference.
- [x] Sign the Contributor License Agreement as prompted for your first-time contribution(*One-time setup*).
- [x] Learn the [coding contribution guide](https://cloudberrydb.org/contribute/code), including our code conventions, workflow and more.
- [x] List your communication in the [GitHub Issues](https://github.com/cloudberrydb/cloudberrydb/issues) or [Discussions](https://github.com/orgs/cloudberrydb/discussions) (if has or needed).
- [x] Document changes.
- [x] Add tests for the change
- [x] Pass `make installcheck`
- [x] Pass `make -C src/test installcheck-cbdb-parallel`
- [x] Feel free to request `cloudberrydb/dev` team for review and approval when your PR is ready🥳

### CODE REVIEW

**Comment by reshke** [comment]
This change is pure interface substitution, no other features/regression added. There are still some more patch needed to make Yezzey work fine with CBDB, but I considered to split there changes into separate work.

**Review by lss602726449 — COMMENTED** [review]
It is an interesting task to read and write files on S3 through the implementation of Appendonly table's smgr.

File: src/backend/access/aocs/test/aocsam_test.c [hunk 1]
```diff
@@ -19,15 +20,23 @@ test__aocs_begin_headerscan(void **state)
 {
 	AOCSHeaderScanDesc desc;
 	RelationData reldata;
+	SMgrRelationData smgrdata;
 	FormData_pg_class pgclass;
 
+	memset(&reldata, 0, sizeof(SMgrRelationData));
```
[the remark this answers is not in the document]
Comment (reshke) [reply src/backend/access/aocs/test/aocsam_test.c:26]: Ah, yes, i intend to memset reldata, but wrong sizeof()

File: src/include/storage/smgr.h [hunk 2]
```diff
@@ -73,6 +74,8 @@ typedef struct SMgrRelationData
 	char				smgr_relpersistence;
 	/* pointer to storage manager */
 	const struct f_smgr *smgr;
+    /*pointer to AO storage manager */
```
[the remark this answers is not in the document]
Comment (reshke) [reply src/include/storage/smgr.h:77]: Sure

File: src/include/utils/datumstream.h [hunk 3]
```diff
@@ -274,7 +275,7 @@ extern DatumStreamRead *create_datumstreamread(
 					   Form_pg_attribute attr,
 					   char *relname,
 					   char *title,
-					   RelFileNode *relFileNode);
+					   RelFileNode *relFileNode, const struct f_smgr_ao *smgrAO);
```
[the remark this answers is not in the document]
Comment (reshke) [reply src/include/utils/datumstream.h:278]: Sure!

**Comment by reshke** [comment]
> It is an interesting task to read and write files on S3 through the implementation of Appendonly table's smgr.

Thats exactly the point. And this is how Yezzey extension works for GP6.

File: src/include/storage/smgr.h [hunk 4]
```diff
@@ -127,12 +130,25 @@ typedef struct f_smgr
...
+
+
 typedef void (*smgr_init_hook_type) (void);
 typedef void (*smgr_hook_type) (SMgrRelation reln, BackendId backend, SMgrImpl which, Relation rel);
 typedef void (*smgr_shutdown_hook_type) (void);
+typedef void (*smgrao_hook_type)(SMgrRelation reln, BackendId backend, SMgrImpl which, Relation rel);
 extern PGDLLIMPORT smgr_init_hook_type smgr_init_hook;
 extern PGDLLIMPORT smgr_hook_type smgr_hook;
 extern PGDLLIMPORT smgr_shutdown_hook_type smgr_shutdown_hook;
+extern PGDLLIMPORT smgrao_hook_type smgrao_hook;
```
[the remark this answers is not in the document]
Comment (reshke) [reply src/include/storage/smgr.h]: Indeed, removed.

**Review by lss602726449 — APPROVED** [review]

**Review by my-ship-it — APPROVED** [review]

**Comment by my-ship-it** [comment]
Please resolve conflicts

**Comment by reshke** [comment]
> Please resolve conflicts

Hi!
Conflicts resolved !

**Comment by tuhaihe** [comment]
Hi @my-ship-it It would be better to reapprove this PR because there have been code changes since your last approval.

**Comment by reshke** [comment]
Any updates?

**Comment by reshke** [comment]
@my-ship-it Hi! Can you provide any review on this PR please?

**Review by my-ship-it — APPROVED** [review]

**Pull request merged**